### PR TITLE
Added twig extension which allows to trigger twig profiles

### PR DESCRIPTION
**Pull request by OliverSkroblin**
Since the YIELD change in Twig, the different blocks are no longer displayed in the various profilers, making it difficult to track which parts are consuming significant rendering time. That's why I implemented this function, which triggers the already existing profiler extension from Shopware and forwards this debug information to the configured profiler integrations.

## Symfony profiler bar
image

## Tideways
image

### CODE REVIEW

**Comment by OliverSkroblin** [comment]
I'm not sure if creating so many traces might cause performance issues or other problems. With single traces, there were no noticeable issues.

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig [hunk 1]
```diff
@@ -98,6 +100,8 @@
                                         {% endblock %}
                                     </div>
                                 {% endfor %}
+                                {{ profiler_start('product-boxes') }}
+
```
Comment (aragon999) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
                                {{ profiler_end('product-boxes') }}
```

:thinking:

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig [hunk 2]
```diff
@@ -135,3 +139,5 @@
         {% endblock %}
     </div>
 {% endblock %}
+
+{{ profiler_start('product-listing') }}
```
Comment (aragon999) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
{{ profiler_end('product-listing') }}
```

**Review by aragon999 — COMMENTED** [review]
Not sure, but I think it would be more reasonable to use the Twig statement thingies (not sure how they are called), i.e. `{% ... %}` and not the Twig output thingies `{{ ... }}`, or is there a specific reason to not use them?

**Comment by Gaitholabi** [comment]
How does the traces look like without the twig extension functions? 
it feels like wrapping the `loadTemplate` is doing the heavy-lifting and is mostly sufficient.

**Comment by OliverSkroblin** [comment]
> How does the traces look like without the twig extension functions? it feels like wrapping the `loadTemplate` is doing the heavy-lifting and is mostly sufficient.

Yes, that's true. The only question is whether having so many profiles might cause problems later on.

**Comment by OliverSkroblin** [comment]
> How does the traces look like without the twig extension functions? it feels like wrapping the `loadTemplate` is doing the heavy-lifting and is mostly sufficient.

image

**Comment by OliverSkroblin** [comment]
Final result

image

image

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig (hunk 2 above)
Comment (OliverSkroblin) [reply src/Storefront/Resources/views/storefront/component/product/listing.html.twig]: fixed

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig (hunk 1 above)
Comment (OliverSkroblin) [reply src/Storefront/Resources/views/storefront/component/product/listing.html.twig]: fixed

**Comment by Gaitholabi** [comment]
> Final result
> 
> image image



Thanks, the result looks really good. Aside from potential performance issues, there's the problem of having a lot of profiling calls scattered across all the templates.

Would there potentially be another solution, where those invocations are added on the fly instead of adding them manually to the templates?

**Comment by OliverSkroblin** [comment]
In my opinion, the solution is now equivalent to the PHP solution. By default, the profilers should automatically detect the corresponding entry points.

These manual profiler calls (as they also exist in PHP) are simply additional aids for "grouping" and "orientation" that developers can add if needed.

Since I have now removed the generic "loadTemplate" wrapper, there should no longer be any performance issues.

**Comment by Gaitholabi** [comment]
> In my opinion, the solution is now equivalent to the PHP solution. By default, the profilers should automatically detect the corresponding entry points.
> 
> These manual profiler calls (as they also exist in PHP) are simply additional aids for "grouping" and "orientation" that developers can add if needed.
> 
> Since I have now removed the generic "loadTemplate" wrapper, there should no longer be any performance issues.

We agree that grouping can indeed be helpful; however, we still find modifying all the Twig templates somewhat intrusive, as it may not provide enough added value to justify these changes.

**Comment by OliverSkroblin** [comment]
> We agree that grouping can indeed be helpful; however, we still find modifying all the Twig templates somewhat intrusive, as it may not provide enough added value to justify these changes.

Alright... I don’t really understand why setting these traces in Twig is an issue, while it’s not in PHP... but okay.

So, can I expect that you’ll provide a solution to regain insights into Twig performance? Because I won’t be putting any more effort into this.

**Comment by Gaitholabi** [comment]
> > In my opinion, the solution is now equivalent to the PHP solution. By default, the profilers should automatically detect the corresponding entry points.
> > These manual profiler calls (as they also exist in PHP) are simply additional aids for "grouping" and "orientation" that developers can add if needed.
> > Since I have now removed the generic "loadTemplate" wrapper, there should no longer be any performance issues.
> 
> We agree that grouping can indeed be helpful; however, we still find modifying all the Twig templates somewhat intrusive, as it may not provide enough added value to justify these changes.

In my personal view, they are also intrusive in PHP and not ideal. 

We believe that wrapping the loadTemplate function should be sufficient. Performance wouldn't be an issue if it's only enabled in the debug environment. We'd still be losing the level of detail that the group per 'block' provides though.

**Comment by BrocksiNet** [comment]
So, can we close this in favor of [GcProfiler](https://github.com/Goblin-coders/GcProfiler)?

**Comment by OliverSkroblin** [comment]
I closed it ;). Will backport this to other versions via plugin

**Pull request closed without merging**